src/coreclr/inc/shash.h

```diff
@@ -71,9 +71,15 @@
 //                                              default traits if it can be assigned from 0.
 // static const bool IsDeleted(const ELEMENT &e) Compare element with Deleted sentinal value. May be inherited from the
 //                                              default traits if it can be assigned from -1.
+// static bool ShouldDelete(const ELEMENT &e)   Called in addition to IsDeleted() when s_supports_autoremove is true, see more
+//                                              information there.
 //
 // static void OnDestructPerEntryCleanupAction(ELEMENT& e) Called on every element when in hashtable destructor.
 //                                              s_DestructPerEntryCleanupAction must be set to true if implemented.
+// static void OnRemovePerEntryCleanupAction(ELEMENT& e) Called when an element is removed from the hashtable.
+//                                              s_RemovePerEntryCleanupAction must be set to true if implemented. When elements
+//                                              are removed in the destructor, OnDestructPerEntryCleanupAction() is called
+//                                              instead if s_DestructPerEntryCleanupAction is true.
```

Comment:
Rather than having two different cleanup operations could we have a single cleanup operation with a boolean flag that says whether destruct also runs in the Remove() function? Having two of them makes it harder to know which one will run in which circumstances. I didn't notice that we have any scenario that depends on having both of the cleanup actions set simultaneously so it feels like this supports more options than our code needs.

Reply:
Looking at the existing uses of OnDestruct, they look like they would work just fine with the newer OnRemove model, and in fact are not really designed for the OnDestruct scenario. I'm not convinced we should keep  OnDestruct at all.